python-avd/pyavd/_schema/models/avd_model.py

```diff
@@ -0,0 +1,341 @@
+
+        If there is no default value in the field info, we return the default-default depending on type.
+        - For lists and dicts we return new empty list / dict.
+        - For AvdModel subclasses we return a new empty instance of the class.
+        - For other types we return None.
+        """
+        if name not in cls._fields:
+            raise AttributeError("'" + cls.__name__ + "' object has no attribute '" + name + "'")
+        field_info = cls._fields[name]
+        field_type: type = field_info["type"]
+        if field_type is list:
+            if (default_value_or_function := field_info.get("default")) is None:
+                return []
+            if isinstance(default_value_or_function, list):
+                # Create a new list
+                return default_value_or_function.copy()
+
+            # Callable default value.
+            return default_value_or_function(field_info["items"])
+
+        if issubclass(field_type, AvdBase) or field_type is dict:
+            return default_function(field_type) if (default_function := field_info.get("default")) else field_type()
+
+        return field_info.get("default")
+
+    def __getattr__(self, name: str) -> Any:
+        """
+        Resolves the default value for a field, set the default value on the attribute and return the value.
+
+        We only get here if the attribute is not set already, and next call will skip this since the attribute is set.
+        """
+        if name not in self._fields:
+            raise AttributeError("'" + self.__class__.__name__ + "' object has no attribute '" + name + "'")
+
+        new_default_value = self._get_field_default_value(name)
+        setattr(self, name, new_default_value)
+        return new_default_value
+
+    def _get_defined_attr(self, name: str) -> Any | Undefined:
+        """
+        Get attribute or Undefined.
+
+        Avoids the overridden __getattr__ to avoid default values.
+        """
+        if name not in self._fields:
+            raise AttributeError("'" + self.__class__.__name__ + "' object has no attribute '" + name + "'")
+        try:
+            return self.__getattribute__(name)
+        except AttributeError:
+            return Undefined
+
+    def __repr__(self) -> str:
+        """Returns a repr with all the fields that a set including any nested models."""
+        cls_name = self.__class__.__name__
+        attrs = [f"{key}={getattr(self, key)!r}" for key in (self._fields or ()) if self._get_defined_attr(key) is not Undefined]
+        return f"<{cls_name}({', '.join(attrs)})>"
+
+    def __bool__(self) -> bool:
+        """
+        Boolean check on the class to quickly determine if any parameter is set.
+
+        The check includes the default values and is performed recursively on any nested models.
+        """
+        return any(getattr(self, key, None) for key in (self._fields or ()) if self._get_defined_attr(key) is not Undefined)
```

Comment:
why `(self._fields or ())` ? can _fields be something else than a dict?

Reply:
I think I had a case where fields where unset and at the time I did not have the overload of __getattr__. I will keep the comment open and investigate later.

Reply:
removed